### ISSUE TITLE
JsonPatch escape hatches are applied twice in CI

### ISSUE THREAD

**Issue by christophgysin** [cdklabs/cdk-pipelines-github#380]
[YamlFile.writeFile()](https://github.com/cdklabs/cdk-pipelines-github/blob/main/src/yaml-file.ts#L137) calls [YamlFile.toYaml()](https://github.com/cdklabs/cdk-pipelines-github/blob/main/src/yaml-file.ts#L121) which calls [JsonPatch.apply()](https://github.com/cdklabs/cdk-pipelines-github/blob/main/src/json-patch.ts#L29). There,  patching is done with [applyPatch() from fast-json-patch](https://github.com/Starcounter-Jack/JSON-Patch#function-applypatchtdocument-t-patch-operation-validateoperation-boolean--validatort-mutatedocument-boolean--true-banprototypemodifications-boolean--true-patchresultt), which modifies the passed object.

In CI, we first call [YamlFile.toYaml()](https://github.com/cdklabs/cdk-pipelines-github/blob/main/src/pipeline.ts#L303) in the diffProtection block, followed by [YamlFile.writeFile()](https://github.com/cdklabs/cdk-pipelines-github/blob/main/src/pipeline.ts#L308).

This causes the patches to be applied twice. That works only if the patches are idempotent. If you define a patch that removes a list element:
```ts
pipeline.workflowFile.patch(JsonPatch.remove('/jobs/myJob/steps/0'))
```
The first element will be remove once locally, but twice in CI.

To fix this, we could:
- change YamlFile.toYaml() to clone the object before passing it to JsonPatch.apply()
- change JsonPatch.apply() to clone the document before applying patches
- change GitHubWorkflow.doBuildPipeline() to skip calling YamlFile.writeFile() in CI

**Comment by kaizencc** [comment]
Huh. Good catch. I think we want `toYaml` to be idempotent. That makes me feel like option 1 is the best: change `YamlFile.toYaml()` to clone the object before passing it to JsonPatch.apply(). Are you willing to submit a PR for this @christophgysin?

**Comment by christophgysin** [comment]
@kaizencc Sure, please check my proposed fix in PR #384
